### PR TITLE
[SPARK-48486][SQL] To solve the issue of generating excessively large execution plans when encountering multiple levels of subqueries while enabling DynamicPartitionPruning.

### PR DESCRIPTION
**Pull request by chenfengbin2009**
[SPARK-48486][CORE] Instead of PartitionPruning#prune is expected
### What changes were proposed in this pull request?

The code of method: prune in class: org.apache.spark.sql.execution.dynamicpruning.PartitionPruning is modified. And class: org.apache.spark.sql.internal.SQLConf has added a property value: DYNAMIC_PARTITION_PRUNING_MAX_LENGTH and a dynamicPartitionPruningMaxLength method.

### Why are the changes needed?

In order to address the issue of Spark's DynamicPartitionPruning generating plans with excessive character length and causing excessive memory usage leading to OOM errors when encountering too many levels of subqueries.

### Does this PR introduce _any_ user-facing change?
yes
When DYNAMIC_PARTITION_PRUNING_ENABLED is set to true and DPP optimization is enabled, the generated execution plan may trigger a potential issue.

### How was this patch tested?
I have provided test cases as much as possible.

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Mind filing a JIRA please? See also https://spark.apache.org/contributing.html

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -351,6 +351,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val DYNAMIC_PARTITION_PRUNING_MAX_LENGTH =
+    buildConf("spark.sql.optimizer.dynamicPartitionPruning.max.length")
+      .doc("The maximum character length of the plan for the nested subquery.When the set length " +
+        "is too large, it may cause the final plan to exceed the limit or memory overflow.")
+      .version("3.3.0")
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:358]: Let's make it 4.0.0

Comment (chenfengbin2009) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:358]: Your suggestion is very good. I will modify the relevant content and submit it again. Do you have any better suggestions for the modified code?

Comment (chenfengbin2009) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: The adapted version has been changed from 3.3.0 to 4.0.0.

**Comment by nikhil-zlai** [comment]
any updates on this issue?

**Pull request closed without merging**